File: app/src/main/java/cat/xojan/random1/viewmodel/PodcastsViewModel.java
package cat.xojan.random1.viewmodel;

import java.util.List;

import javax.inject.Inject;

import cat.xojan.random1.domain.entities.Podcast;
import cat.xojan.random1.domain.entities.Program;
import cat.xojan.random1.domain.entities.Section;
import cat.xojan.random1.domain.interactor.ProgramDataInteractor;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.subjects.PublishSubject;

public class PodcastsViewModel {

    private final ProgramDataInteractor mProgramDataInteractor;

    @Inject
    public PodcastsViewModel(ProgramDataInteractor programDataInteractor) {
        mProgramDataInteractor = programDataInteractor;
    }

    public Single<List<Podcast>> loadDownloadedPodcasts() {
        return mProgramDataInteractor.getDownloadedPodcasts();
    }

    public PublishSubject<List<Podcast>> getDownloadedPodcastsUpdates() {
        return mProgramDataInteractor.getDownloadedPodcastsUpdates();
    }

    public Single<List<Podcast>> loadPodcasts(Program program, Section section,
                                                  boolean refresh) {
        Single<List<Podcast>> loadedPodcasts =
                mProgramDataInteractor.loadPodcasts(program, section, refresh)
                .flatMapIterable(list -> list)
                .map(podcast -> {
                    podcast.setProgramId(program.getId());
                    podcast.setImageUrl(program.getImageUrl());
                    return podcast;
                })
                .toList();

        Single<List<Podcast>> downloadedPodcasts =
                mProgramDataInteractor.getDownloadedPodcasts();

        return Single.zip(loadedPodcasts, downloadedPodcasts,
                (loaded, downloaded) -> {
                    for (Podcast podcast : loaded) {
                        podcast.setFilePath(null);
                        podcast.setState(Podcast.State.LOADED);
                    }

                    for (Podcast download : downloaded) {
                        int index = loaded.indexOf(download);
                        if (index >= 0) {
                            Podcast podcast = loaded.get(index);
                            podcast.setFilePath(download.getFilePath());
                            podcast.setState(download.getState());
                        }
                    }
                    return loaded;
                });
    }

    public void selectedSection(boolean b) {
        mProgramDataInteractor.setSectionSelected(b);
    }

    public Observable<Boolean> exportPodcasts() {
        return mProgramDataInteractor.exportPodcasts();
    }
}

File: app/src/main/java/cat/xojan/random1/viewmodel/PodcastsViewModel.kt
package cat.xojan.random1.viewmodel

import javax.inject.Inject

import cat.xojan.random1.domain.entities.Podcast
import cat.xojan.random1.domain.entities.Program
import cat.xojan.random1.domain.entities.Section
import cat.xojan.random1.domain.interactor.ProgramDataInteractor
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.functions.BiFunction
import io.reactivex.subjects.PublishSubject

class PodcastsViewModel @Inject
constructor(private val mProgramDataInteractor: ProgramDataInteractor) {

    val downloadedPodcastsUpdates: PublishSubject<List<Podcast>>
        get() = mProgramDataInteractor.getDownloadedPodcastsUpdates()

    fun loadDownloadedPodcasts(): Single<List<Podcast>> {
        return mProgramDataInteractor.getDownloadedPodcasts()
    }

    fun loadPodcasts(program: Program, section: Section,
                     refresh: Boolean): Single<List<Podcast>> {
        val loadedPodcasts = mProgramDataInteractor.loadPodcasts(program, section, refresh)
                .flatMapIterable { list -> list }
                .map { podcast ->
                    podcast.programId = program.id
                    podcast.imageUrl = program.imageUrl()
                    podcast
                }
                .toList()

        val downloadedPodcasts = mProgramDataInteractor.getDownloadedPodcasts()

        return Single.zip(loadedPodcasts, downloadedPodcasts, BiFunction {
            loaded, downloaded ->
            for (podcast in loaded) {
                podcast.filePath = null
                podcast.state = Podcast.State.LOADED
            }

            for (download in downloaded) {
                val index = loaded.indexOf(download)
                if (index >= 0) {
                    val podcast = loaded.get(index)
                    podcast.filePath = download.filePath
                    podcast.state = download.state
                }
            }
            loaded
        })
    }

    fun selectedSection(b: Boolean) {
        mProgramDataInteractor.setSectionSelected(b)
    }

    fun exportPodcasts(): Observable<Boolean> {
        return mProgramDataInteractor.exportPodcasts()
    }
}
